### PR TITLE
[22188]Set allowed timestamp skew

### PR DESCRIPTION
**Pull request by reuvenlax**

### CODE REVIEW

**Comment by reuvenlax** [comment]
R: @pabloem

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by johnjcasey** [comment]
Run Java PreCommit

**Comment by johnjcasey** [comment]
@reuvenlax why did you pick 1 year?

**Comment by kennknowles** [comment]
R: @johnjcasey

**Comment by kennknowles** [comment]
Yea, I don't see this as the solution - we have a problem where an illegal timestamp is being created and no amount of allowed lateness would change that issue.

**Comment by johnjcasey** [comment]
My understanding is that this is part 2 of a two part solution to the problem. @reuvenlax can you confirm that?

**Comment by reuvenlax** [comment]
This is part 2. I added a comment to address Kenn's concern, but I still
haven't been able to push changes.

**Comment by reuvenlax** [comment]
@kennknowles comment was correct. The intention here was to set allowedTimestampSkew, not allowed lateness. The original PR was incorrect, but this is now fixed.

**Comment by johnjcasey** [comment]
is this intended to delete playground files?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java [hunk 1]
```diff
@@ -553,6 +553,11 @@ public void onTimer(
         OutputReceiver<KV<String, Operation>> o,
         BoundedWindow window) {
       // Stream is idle - clear it.
+      // Note: this is best effort. We are explicitly emiting a timestamp that is before
+      // the default output timestamp, which means that in some cases (usually when draining
+      // a pipeline) this finalize element will be dropped as late. This is usually ok as
+      // BigQuery will eventually garbage collect the stream. We attempt to finalize idle streams
+      // merely to remove the pressure of large numbers of orphaned streams from BigQuery.
```
Comment (johnjcasey) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:560]: what is the consequence of orphaned streams?

**Comment by kennknowles** [comment]
The timestamp in #22188 is beyond the maximum timestamp allowed in Beam, so adding allowed skew cannot be the full solution.

**Comment by reuvenlax** [comment]
@kennknowles Actually setting the allowed skew to be 1ms would probably be sufficient here. I just set it to the max. Also see comment - there are definitely some scenarios where the output will still be dropped as late, but that is ok as this is a best-effort cleanup to take pressure off of BigQuery

**Comment by kennknowles** [comment]
But the timestamp is outside of Beam's allowed range of timestamps. It is just invalid data, right? Even if there is a short term change that will get around the current check, it is still bad data.

**Comment by reuvenlax** [comment]
The time on the production is always window.maxTimestamp(), which should be
in the valid range.

**Comment by kennknowles** [comment]
I think that creation of the window must not be validated, or some such. `294247-01-10T04:00:54.776Z` is one milli past the maximum allowed.

**Comment by reuvenlax** [comment]
Ah I see what the confusion is.

This timer is set using the withNoOutputTimestamp, largely because
otherwise the watermark hold causes problems. That was the source of the
original bug, because we shouldn't be outputting from a timer set like
that. A timer set using withNoOutputTimestamp has the default output
timestamp set to an invalid value as you noticed.

This fix is a hack - explicitly setting the production timestamp to a valid
value (end of window) and then overriding the allowed timestamp skew to a
value >= 1ms allows the production to go through. However it is sketchy -
it relies on knowing how withNoOutputTimestamp is implemented.

The "correct" fix would require more significant refactoring. Given that
this is a regression that makes the sink unusable for many users, I wanted
to get a fix into the current beam release first and then replace this with
a more-principled fix (which would be to do the cleanup inline in onTimer
instead of delegating to another ParDo).

**Comment by reuvenlax** [comment]
@johnjcasey per the above discussion, we now finalize idle streams inline in onTimer instead of using the FlushAndFinalizeDoFn. Long-term I would like to share this code as it's 98% the same, however the presence of Beam counters throughout makes it hard to share the code.

**Comment by kennknowles** [comment]
Ah I was not familiar with `withNoOutputTimestamp`. But everything has a timestamp so setting it to `window.maxTimestamp()` should be equivalent and less confusing.

**Comment by kennknowles** [comment]
Ah - or if it means that it is a timer that cannot output at all but only manipulate state?

**Comment by reuvenlax** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by reuvenlax** [comment]
@kennknowles could you review since @johnjcasey is out now. I think this needs to be in the current release, as we are fixing a major regression introduced in the last release.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java (hunk 1 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java]: if there are too many of them, BigQuery might start rejecting future stream creation requests

**Comment by reuvenlax** [comment]
@kennknowles correct. In this case the timer now manipulates state and has a side effect (telling BigQuery to finalize the stream).

**Comment by kileys** [comment]
Run Java PreCommit

**Comment by reuvenlax** [comment]
After more thought, this alternate fix is incorrect and can result in duplicates. The previous commit is hacky, but will work today.

**Comment by reuvenlax** [comment]
this is ready

**Comment by johnjcasey** [comment]
@kennknowles should probably review this instead of me. He has better context on this than me. 
I believe this looks good, but I wouldn't have caught most of what Kenn caught on earlier passes

**Comment by aaltay** [comment]
@kennknowles is OOO until 8/12. He would not be able to review this.

R: @chamikaramj would you be able to review it?

**Comment by aaltay** [comment]
Alternatively if this is passing tests, and @kennknowles and @johnjcasey already had a chance to look at it we might merge it.

**Comment by johnjcasey** [comment]
If Kenn is out of office, and this needs to be in the release, we can probably merge this then.

**Comment by reuvenlax** [comment]
While there is some discussion as to what the "right" fix is (the "right"
fix likely requires functionality that is not yet in the Beam model), this
PR is quite minimal and does fix the problem.

**Comment by kileys** [comment]
If we're ok merging this for the release, could you setup a cherry-pick PR for the release?

**Comment by reuvenlax** [comment]
@kileys if you're ok, could you LGTM?

**Comment by kileys** [comment]
I'm not a 100% on this, but it LGTM

**Pull request merged**